File: server/src/collectors/tellusSatellite.js
/**
 * Tellus satellite collector (tellus-satellite) — kind:'intel'.
 *
 * Tellus (さくらインターネット / 経産省) is Japan's national satellite-data
 * platform. Its data API (catalog / scene search) requires an API token.
 * We read TELLUS_TOKEN via envFor (BYOK). When absent we return honest empty
 * ('tellus-satellite_no_key'); when present we query the Tellus data-search
 * API for the latest scenes over Japan and emit one intel item per scene
 * (raster product, not a point cloud).
 *
 * The Tellus Traveler / data-search API path is unverified across versions;
 * any auth/fetch failure yields honest empty ('tellus-satellite_unavailable').
 * Never fabricated.
 */

import { fetchJson } from './_liveHelpers.js';
import { envFor } from '../utils/collectorEnv.js';
import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';

const SOURCE_ID = 'tellus-satellite';
const JP_BBOX = [122.0, 24.0, 146.0, 46.0];

function empty(source, description) {
  return intelEnvelope({
    sourceId: SOURCE_ID,
    items: [],
    live: false,
    description,
    extraMeta: { source },
  });
}

export default async function collectTellusSatellite() {
  const token = envFor('TELLUS_TOKEN');
  if (!token) {
    return empty('tellus-satellite_no_key',
      'Tellus satellite scene catalog over Japan (requires TELLUS_TOKEN)');
  }

  const [w, s, e, n] = JP_BBOX;
  // Tellus Traveler data-search API (STAC-like). Endpoint/route unverified
  // across Tellus API versions; honest empty on any failure.
  const reqBody = {
    intersects: {
      type: 'Polygon',
      coordinates: [[[w, s], [e, s], [e, n], [w, n], [w, s]]],
    },
    limit: 50,
    sortby: [{ field: 'properties.start_datetime', direction: 'desc' }],
  };
  const data = await fetchJson(
    'https://www.tellusxdp.com/api/traveler/v1/data-search/',
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(reqBody),
      timeoutMs: 20000,
    },
  ).catch(() => null);

  const feats = Array.isArray(data?.features) ? data.features
    : Array.isArray(data?.results) ? data.results
      : Array.isArray(data?.items) ? data.items : null;
  if (!feats || feats.length === 0) {
    return empty('tellus-satellite_unavailable',
      'Tellus satellite scene catalog over Japan — API returned no scenes');
  }

  const items = feats.slice(0, 50).map((f, i) => {
    const props = f.properties || f;
    const sid = f.id || props.scene_id || props.id || `scene-${i}`;
    const acquired = props.start_datetime || props.datetime
      || props.acquired || null;
    return {
      uid: intelUid(SOURCE_ID, sid),
      title: `Tellus scene ${sid}`,
      summary: `Tellus satellite scene acquired ${acquired || 'unknown'} over Japan.`,
      body: `Tellus platform satellite scene ${sid} intersecting the Japan AOI. Acquired ${acquired || 'unknown'}. Sensor ${props.sensor || props.platform || 'unknown'}.`,
      link: 'https://www.tellusxdp.com/',
      published_at: acquired || new Date().toISOString(),
      tags: ['satellite', 'tellus', 'japan', 'raster'],
      properties: {
        bbox: f.bbox || JP_BBOX,
        geometry: f.geometry || null,
        acquired,
        sensor: props.sensor || null,
        platform: props.platform || null,
        scene_id: sid,
      },
    };
  });

  return intelEnvelope({
    sourceId: SOURCE_ID,
    items,
    live: true,
    description: 'Latest Tellus platform satellite scenes over Japan',
    extraMeta: { source: 'tellus-satellite' },
  });
}

File: native/collectors/sources/tellus_satellite.c
/* collectors/satellite/sources/tellus_satellite.c
 * Port of server/src/collectors/tellusSatellite.js.
 * Tellus Traveler data-search API (POST, Bearer TELLUS_TOKEN) → one intel
 * item per scene. Non-spatial scene catalog (has_geo=0). Gated on
 * TELLUS_TOKEN (no key-free path). Honest empty on fetch failure. */
#include "../../source.h"
#include "../../lib/feedlib.h"
#include "../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static const char *sstr(cJSON *o, const char *k) {
  cJSON *v = o ? cJSON_GetObjectItem(o, k) : NULL;
  return (v && cJSON_IsString(v) && v->valuestring[0]) ? v->valuestring : NULL;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  const char *token = getenv("TELLUS_TOKEN");
  if (!token || !*token) {
    fprintf(stderr, "[tellus-satellite] gated (no TELLUS_TOKEN)\n");
    return 0;
  }

  /* JP_BBOX = [122,24,146,46] -> intersects polygon */
  const char *body =
    "{\"intersects\":{\"type\":\"Polygon\",\"coordinates\":"
    "[[[122,24],[146,24],[146,46],[122,46],[122,24]]]},"
    "\"limit\":50,\"sortby\":[{\"field\":\"properties.start_datetime\","
    "\"direction\":\"desc\"}]}";

  char auth[600];
  snprintf(auth, sizeof auth, "Authorization: Bearer %s", token);
  const char *hdrs[] = { "Content-Type: application/json", auth, NULL };

  cJSON *data = feed_post_json(ctx->http,
    "https://www.tellusxdp.com/api/traveler/v1/data-search/",
    body, hdrs, 20000);

  cJSON *feats = NULL;
  if (data) {
    cJSON *f;
    if ((f = cJSON_GetObjectItem(data, "features")) && cJSON_IsArray(f))
      feats = f;
    else if ((f = cJSON_GetObjectItem(data, "results")) && cJSON_IsArray(f))
      feats = f;
    else if ((f = cJSON_GetObjectItem(data, "items")) && cJSON_IsArray(f))
      feats = f;
  }
  if (!feats || cJSON_GetArraySize(feats) == 0) {
    if (data) cJSON_Delete(data);
    fprintf(stderr, "[tellus-satellite] unavailable (no scenes)\n");
    return -1;
  }

  int n = 0, i = 0;
  cJSON *f;
  cJSON_ArrayForEach(f, feats) {
    if (i >= 50) break;
    cJSON *props = cJSON_GetObjectItem(f, "properties");
    if (!props || !cJSON_IsObject(props)) props = f;

    char sidbuf[32];
    const char *sid = sstr(f, "id");
    if (!sid) sid = sstr(props, "scene_id");
    if (!sid) sid = sstr(props, "id");
    if (!sid) { snprintf(sidbuf, sizeof sidbuf, "scene-%d", i); sid = sidbuf; }

    const char *acquired = sstr(props, "start_datetime");
    if (!acquired) acquired = sstr(props, "datetime");
    if (!acquired) acquired = sstr(props, "acquired");
    const char *sensor = sstr(props, "sensor");
    const char *platform = sstr(props, "platform");

    char title[128], summary[256], bodytxt[512];
    snprintf(title, sizeof title, "Tellus scene %s", sid);
    snprintf(summary, sizeof summary,
      "Tellus satellite scene acquired %s over Japan.",
      acquired ? acquired : "unknown");
    snprintf(bodytxt, sizeof bodytxt,
      "Tellus platform satellite scene %s intersecting the Japan AOI. "
      "Acquired %s. Sensor %s.",
      sid, acquired ? acquired : "unknown",
      sensor ? sensor : (platform ? platform : "unknown"));

    cJSON *p = cJSON_CreateObject();
    cJSON *bb = cJSON_GetObjectItem(f, "bbox");
    if (bb) cJSON_AddItemToObject(p, "bbox", cJSON_Duplicate(bb, 1));
    else {
      cJSON *jb = cJSON_CreateArray();
      cJSON_AddItemToArray(jb, cJSON_CreateNumber(122.0));
      cJSON_AddItemToArray(jb, cJSON_CreateNumber(24.0));
      cJSON_AddItemToArray(jb, cJSON_CreateNumber(146.0));
      cJSON_AddItemToArray(jb, cJSON_CreateNumber(46.0));
      cJSON_AddItemToObject(p, "bbox", jb);
    }
    cJSON *gm = cJSON_GetObjectItem(f, "geometry");
    cJSON_AddItemToObject(p, "geometry",
      gm ? cJSON_Duplicate(gm, 1) : cJSON_CreateNull());
    if (acquired) cJSON_AddStringToObject(p, "acquired", acquired);
    else cJSON_AddNullToObject(p, "acquired");
    if (sensor) cJSON_AddStringToObject(p, "sensor", sensor);
    else cJSON_AddNullToObject(p, "sensor");
    if (platform) cJSON_AddStringToObject(p, "platform", platform);
    else cJSON_AddNullToObject(p, "platform");
    cJSON_AddStringToObject(p, "scene_id", sid);
    char *pj = cJSON_PrintUnformatted(p);

    cJSON *tags = cJSON_CreateArray();
    cJSON_AddItemToArray(tags, cJSON_CreateString("satellite"));
    cJSON_AddItemToArray(tags, cJSON_CreateString("tellus"));
    cJSON_AddItemToArray(tags, cJSON_CreateString("japan"));
    cJSON_AddItemToArray(tags, cJSON_CreateString("raster"));
    char *tj = cJSON_PrintUnformatted(tags);

    intel_item it = {0};
    it.remote_key = sid;             /* uid tellus-satellite|<sid> */
    it.title = title;
    it.summary = summary;
    it.body = bodytxt;
    it.link = "https://www.tellusxdp.com/";
    it.published_at = acquired;
    it.record_type = "tellus-satellite";
    it.has_geo = 0;
    it.properties_json = pj;
    it.tags_json = tj;
    if (sink->emit(sink, &it) >= 0) n++;

    free(pj); free(tj);
    cJSON_Delete(p); cJSON_Delete(tags);
    i++;
  }
  cJSON_Delete(data);
  fprintf(stderr, "[tellus-satellite] emitted %d\n", n);
  return n > 0 ? 0 : -1;
}

static const source_def tellus_satellite_def = {
  .id = "tellus-satellite", .collector = "satellite",
  .name = "Tellus Satellite Imagery", .name_ja = "Tellus 衛星画像",
  .update_interval_sec = 86400, .run = run };
REGISTER_SOURCE(tellus_satellite_def)
